MAINT: stats.mode: fix issues with `nan_policy='propagate'`

mdhaber opened the pull request:
#### Reference issue
closes gh-9815

#### What does this implement/fix?
Two issues were observed in gh-9815:
* `mode` with `nan_policy='propagate'` did not always produce the expected result when the input contained NaNs
* `mode` with `nan_policy='propagate'` encountered an error when the input was an object array containing types that could not be compared

This fixes these issues by replacing `mode`'s fall-back algorithm.

Now:
```python3
import numpy as np
from scipy import stats
stats.mode([np.nan, np.nan, np.nan, 1., 1., 2., 3., 3., 3., np.nan])  # ModeResult(mode=nan, count=4)
stats.mode(np.array([10, True, 'hello', 10], dtype='object'))  # ModeResult(mode=10, count=2)
```

#### Additional information
There are multiple binary representations of NaN. This PR does not attempt to treat all of them as the same value. However, it is careful to preserve the NaN representations of the user input so that they are compared as intended. This is important, because NumPy may alter the representation of NaNs in a sequence when converting it to a numerical array.

<details>

```python3
from collections import Counter
import numpy as np

a = [2, np.nan, 1, np.nan]
Counter(a)  # Counter({2: 1, nan: 2, 1: 1})
Counter(np.asarray(a, dtype=object))  #  Counter({2: 1, nan: 2, 1: 1})
Counter(np.asarray(a, dtype=float))  # Counter({2.0: 1, nan: 1, 1.0: 1, nan: 1})
```

</details>

It might be possible to treat different NaNs as the same, but we would probably have to change them to a sentinel value first (e.g. `a[np.isnan(a)] = 'a_sentinel_value_not_already_in_data'`). For object arrays, we'd have to loop through element-by-element (not just slice by slice) because `np.isnan` doesn't work on object arrays.  Do we want to do that? There was not consensus in gh-9815 on whether that was important or not, just that it was important to document. Either way, we'll explain it in a Notes section.

tirthasheshpatel commented:
> NumPy may alter the representation of NaNs in a sequence when converting it to a numerical array.

Is it documented anywhere?

> It might be possible to treat different NaNs as the same, but we would probably have to change them to a sentinel value first

I think treating all `nan`s the same way would lead to a more intuitive behavior (at least, from the perspective of the user). For example, if two `nan`s were produced in a different manner, they might have different IEEE 784 representations and lead to unintuitive results:

```python
>>> import numpy as np
>>> from scipy.stats import mode
>>> import struct
>>> def binary(num):
...     return ''.join('{:0>8b}'.format(c) for c in struct.pack('!f', num))
>>> zero = np.float64(0)
>>> my_nan = (zero / zero)
<stdin>:1: RuntimeWarning: invalid value encountered in double_scalars
>>> binary(my_nan)
'11111111110000000000000000000000'
>>> binary(my_nan) == binary(np.nan)
False
>>> a = [2, np.nan, 1, my_nan]
>>> x = np.asarray(a, dtype=object)
>>> y = np.asarray(a, dtype=float)
>>> binary(x[1]), binary(x[3])
('01111111110000000000000000000000', '11111111110000000000000000000000')
>>> binary(y[1]), binary(y[3])  # NumPy didn't change the binary representation when converting from object -> float
('01111111110000000000000000000000', '11111111110000000000000000000000')
>>> binary(np.nan)
'01111111110000000000000000000000'
>>> mode(x)
ModeResult(mode=2, count=1)  # oops, mode should be `nan` intuitively
```

tupui reviewed (commented):
The code is clean, we just need to decide on the NaNs as you said.

> I think treating all `nan`s the same way would lead to a more intuitive behavior (at least, from the perspective of the user). For example, if two `nan`s were produced in a different manner, they might have different IEEE 784 representations and lead to unintuitive results:

Are you saying we should use the binary representation to catch all kind of nan? It seems quite difficult to guarantee anything here no?

tirthasheshpatel commented:
> For object arrays, we'd have to loop through element-by-element (not just slice by slice) because `np.isnan` doesn't work on object arrays. Do we want to do that?

Should we treat `nan` specially for object arrays? I think if the user passes an object array, each element should be treated as an object (even all types of `nan`s) and calculations must be done accordingly. To me, `nan_policy` makes sense only for numeric array since `nan`s can be problematic for numeric computations.


> Are you saying we should use the binary representation to catch all kind of nan? It seems quite difficult to guarantee anything here no?

I just mean replacing the nan values with a common `nan` value (or a sentinel value as @mdhaber suggested). Something like `a[np.isnan(a)] = np.nan`. This will convert all the `nan` values to a common representation.

tirthasheshpatel commented:
> Are you saying we should use the binary representation to catch all kind of nan? It seems quite difficult to guarantee anything here no?

@tupui I just mean replacing the nan values with a common `nan` value (or a sentinel value as @mdhaber suggested). Something like `a[np.isnan(a)] = np.nan`. This will convert all the `nan` values to a common representation.

mdhaber commented:
> This should convert all the nan values to a common representation.

I tried that. I don't think it did.

tirthasheshpatel commented:
> I tried that. I don't think it did. This is part of the problem. NumPy's behavior regarding NaNs is weird.

Why not keep using `np.unique` since it treats different types of nans as equivalent and gives expected output:

```python
>>> import numpy as np
>>> zero = np.float64(0)
>>> my_nan = zero / zero
<stdin>:1: RuntimeWarning: invalid value encountered in double_scalars
>>> a = [1, np.nan, 2, my_nan]  # different types of nans
>>> np.unique(x, return_counts=True)
(array([ 1.,  2., nan]), array([1, 1, 2]))
```

> So here's what I think I'm hearing:
For numerical arrays, treat all NaNs as the same.
For object arrays, treat NaNs as objects - different representation NaNs are different.
Is that right?

Yes, that'd simplify things and lead to intuitive behavior.

mdhaber commented:
> Why not keep using np.unique since it treats different types of nans as equivalent and gives expected output:

I think it is either version or machine dependent. That was the original plan (see mdhaber/scipy#69). I switched to my Mac, ran tests before submitting the PR, and got failures. That's why I added `if contains_nan: <convert to object array>`. But on this machine, it's working fine when leaving them as floats.

I'll try leaving them as floats and seeing if tests pass on CI.

mdhaber commented:
Yeah, it seems that `np.unique` is not consistent across platforms/versions. 

Here's the problem, stated concisely:
* When an input sequence is converted to a numerical array, NaNs in the input with identical representations can turn into NaNs with different representations
* On some platforms/versions, `np.unique` treats NaNs with different representations as distinct.

My fix was to convert the input to an object array.
* When the input sequence is converted to an object array, the NaNs maintain their original representation. _This is the heart of the fix: NaNs that start with identical representations stay identical._
* Even though `Counter` treats different NaN representations as distinct, all the NaNs in the input of the tests have the same representation, so they are considered to be the same.

If we really want this to treat all NaNs as identical on all platforms/NumPy versions, I think we have to use a sentinel value.

tirthasheshpatel commented:
> If we really want `mode` to treat all NaNs as identical on all platforms/NumPy versions, I think we have to replace NaNs with a sentinel value. Do we want to go that route?

You noted that for object arrays, we need to go element by element which sounds pretty bad for large arrays. Do you think we can replace all `nan` values with a sentinel and then run `np.unique`? This way we can just avoid coercing the inputs to object arrays.

mdhaber commented:
For numerical input, that's the plan. For general object arrays passed as inputs, no. `mode` has supported object arrays in the past, so we're stuck with them unless we want to deprecate that use.

mdhaber commented:
@tirthasheshpatel @tupui I looked into this a bit and there was indeed [a change in `np.unique` behavior](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/DBJZNCNBXTNU2EWKDZMP7MDQLOC46GNV/?sort=thread) between 1.20 and 1.21 that explains the test failures in the current PR.
I'd suggest we:

- Keep the logic as-is. It sounds like the new behavior (the one we agree with) will [continue to be the default](https://github.com/numpy/numpy/issues/20326).
- Document the difference in behavior depending on NumPy version
- Check NumPy version in the test and xfail for old NumPy
- Merge this
- Discuss deprecation of non-numerical array support separately

How does that sound?

tupui commented:
Sounds good to me, as you showed, it look like NumPy discussed this in length so it would make sense to follow their conclusions and rely on their default.

mdhaber commented:
How about you, @tirthasheshpatel? If you agree, I think this one is ready.

tirthasheshpatel commented:
Thanks for digging into NumPy's changes and issues, @mdhaber!

> How does that sound?

Sounds good to me too!

> you're welcome to open an issue about deprecating non-numerical array support for `mode`.

Done: #15551

tirthasheshpatel reviewed (approved):
The changes LGTM now! We can continue discussing the deprecation of object arrays and adding `_axis_nan_policy_factory` in #15551. Thanks @mdhaber, @tupui, @Kai-Striega!
